internal/asm/c128/asm_test.go

```diff
@@ -0,0 +1,127 @@
+const testLen = 1e5
+
+var x = make([]complex128, testLen)
+
+// guardVector copies the source vector (vec) into a new slice with guards.
+// Guards guarded[:gdLn] and guarded[len-gdLn:] will be filled with sigil value gdVal.
+func guardVector(vec []complex128, gdVal complex128, gdLn int) (guarded []complex128) {
+	guarded = make([]complex128, len(vec)+gdLn*2)
+	copy(guarded[gdLn:], vec)
+	for i := 0; i < gdLn; i++ {
+		guarded[i] = gdVal
+		guarded[len(guarded)-1-i] = gdVal
+	}
+	return guarded
+}
+
+// isValidGuard will test for violated guards, generated by guardVector.
+func isValidGuard(vec []complex128, gdVal complex128, gdLn int) bool {
+	for i := 0; i < gdLn; i++ {
+		if !sameCmplx(vec[i], gdVal) || !sameCmplx(vec[len(vec)-1-i], gdVal) {
+			return false
+		}
+	}
+	return true
+}
+
+// guardIncVector copies the source vector (vec) into a new incremented slice with guards.
+// End guards will be length gdLen.
+// Internal and end guards will be filled with sigil value gdVal.
+func guardIncVector(vec []complex128, gdVal complex128, inc, gdLen int) (guarded []complex128) {
+	if inc < 0 {
+		inc = -inc
+	}
+	inrLen := len(vec) * inc
+	guarded = make([]complex128, inrLen+gdLen*2)
+	for i := range guarded {
+		guarded[i] = gdVal
+	}
+	for i, v := range vec {
+		guarded[gdLen+i*inc] = v
+	}
+	return guarded
+}
+
+// checkValidIncGuard will test for violated guards, generated by guardIncVector
+func checkValidIncGuard(t *testing.T, vec []complex128, gdVal complex128, inc, gdLen int) {
+	srcLn := len(vec) - 2*gdLen
+	for i := range vec {
+		switch {
+		case sameCmplx(vec[i], gdVal):
+			// Correct value
+		case (i-gdLen)%inc == 0 && (i-gdLen)/inc < len(vec):
+			// Ignore input values
+		case i < gdLen:
+			t.Errorf("Front guard violated at %d %v", i, vec[:gdLen])
+		case i > gdLen+srcLn:
+			t.Errorf("Back guard violated at %d %v", i-gdLen-srcLn, vec[gdLen+srcLn:])
+		default:
+			t.Errorf("Internal guard violated at %d %v", i-gdLen, vec[gdLen:gdLen+srcLn])
+		}
+	}
+}
+
+// same tests for nan-aware equality.
+func same(a, b float64) bool {
+	return a == b || (math.IsNaN(a) && math.IsNaN(b))
+}
+
+// sameApprox tests for nan-aware equality within tolerance.
+func sameApprox(a, b, tol float64) bool {
+	return same(a, b) || floats.EqualWithinAbsOrRel(a, b, tol, tol)
+}
+
+// sameCmplx tests for nan-aware equality.
+func sameCmplx(a, b complex128) bool {
+	return a == b || (cmplx.IsNaN(a) && cmplx.IsNaN(b))
+}
+
+// sameCmplxApprox tests for nan-aware equality within tolerance.
+func sameCmplxApprox(a, b complex128, tol float64) bool {
+	return sameCmplx(a, b) || cmplxs.EqualWithinAbsOrRel(a, b, tol, tol)
+}
+
+var ( // Offset sets for testing alignment handling in Unitary assembly functions.
+	align1 = []int{0, 1}
+)
+
+func randomSlice(n, inc int) []complex128 {
+	if inc < 0 {
+		inc = -inc
+	}
+	x := make([]complex128, (n-1)*inc+1)
+	for i := range x {
+		x[i] = complex(rand.Float64(), rand.Float64())
```

Comment:
I think we're trying to avoid rand package-level functions.

Reply:
I'd like to go through and fix up this package and the cognate packages together.